File: apps/firefly/product/client/imp/ios/NoobTest/Toolbar.swift
import SwiftUI

enum ToolbarExplorer {
    case makePost, search, users
}

struct Toolbar: View {
    @Binding var currentExplorer: ToolbarExplorer
    @ObservedObject var tunables = TunableConstants.shared
    let onResetMakePost: () -> Void
    let onResetSearch: () -> Void
    let onResetUsers: () -> Void
    var showPostsBadge: Bool = false
    var showSearchBadge: Bool = false
    var showUsersBadge: Bool = false

    var body: some View {
        HStack(spacing: 0) {
            // Make Post button
            ToolbarButton(icon: "bubble.left", isActive: currentExplorer == .makePost, showBadge: showPostsBadge) {
                if currentExplorer == .makePost {
                    onResetMakePost()
                } else {
                    currentExplorer = .makePost
                    onResetMakePost()  // Also clear badge when switching to posts tab
                }
            }

            Spacer()

            // Search button
            ToolbarButton(icon: "magnifyingglass", isActive: currentExplorer == .search, showBadge: showSearchBadge) {
                if currentExplorer == .search {
                    onResetSearch()
                } else {
                    currentExplorer = .search
                }
            }

            Spacer()

            // Users button
            ToolbarButton(icon: "person.2", isActive: currentExplorer == .users, showBadge: showUsersBadge) {
                if currentExplorer == .users {
                    onResetUsers()
                } else {
                    currentExplorer = .users
                    onResetUsers()  // Also clear badge when switching to users tab
                }
            }
        }
        .padding(.horizontal, 66)  // Increased to move outer buttons inward
        .padding(.vertical, 12)     // Match add post button height
        .background(
            tunables.buttonColor()
                .cornerRadius(12 * tunables.getDouble("corner-roundness", default: 1.0))
                .shadow(color: Color.black.opacity(0.4), radius: 12, x: 0, y: 4)
        )
        .padding(.horizontal, 8 * tunables.getDouble("spacing", default: 1.0))  // Match posts padding
        .offset(y: 34)  // Push down to align bottom edge with screen bottom
        .onAppear {
            // Register toolbar buttons with UI automation
            UIAutomationRegistry.shared.register(id: "toolbar-makepost") {
                currentExplorer = .makePost
            }

            UIAutomationRegistry.shared.register(id: "toolbar-search") {
                currentExplorer = .search
            }

            UIAutomationRegistry.shared.register(id: "toolbar-users") {
                currentExplorer = .users
            }
        }
    }
}

struct ToolbarButton: View {
    let icon: String
    let isActive: Bool
    var showBadge: Bool = false
    let action: () -> Void
    @ObservedObject var tunables = TunableConstants.shared

    var body: some View {
        Button(action: action) {
            Image(systemName: icon)
                .font(.system(size: 20))
                .foregroundColor(.black)
                .frame(width: 35, height: 35)
                .background(
                    isActive ? tunables.buttonHighlightColor() : Color.clear
                )
                .cornerRadius(6)
                .overlay(alignment: .topTrailing) {
                    if showBadge {
                        Circle()
                            .fill(Color.red)
                            .stroke(Color.white, lineWidth: 2)
                            .frame(width: 10, height: 10)
                            .offset(x: 2, y: -2)
                    }
                }
        }
    }
}

File: apps/firefly/product/client/imp/android/app/src/main/kotlin/com/miso/noobtest/Toolbar.kt
package com.miso.noobtest

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.ChatBubbleOutline
import androidx.compose.material.icons.outlined.People
import androidx.compose.material.icons.outlined.Search
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp

enum class ToolbarExplorer {
    MAKE_POST, SEARCH, USERS
}

@Composable
fun Toolbar(
    currentExplorer: ToolbarExplorer,
    onExplorerChange: (ToolbarExplorer) -> Unit,
    onResetMakePost: () -> Unit,
    onResetSearch: () -> Unit,
    onResetUsers: () -> Unit,
    showPostsBadge: Boolean = false,
    showSearchBadge: Boolean = false,
    showUsersBadge: Boolean = false
) {
    // Use rememberUpdatedState to keep the current explorer value up-to-date in callbacks
    val currentExplorerState by rememberUpdatedState(currentExplorer)
    val onExplorerChangeState by rememberUpdatedState(onExplorerChange)
    val onResetMakePostState by rememberUpdatedState(onResetMakePost)
    val onResetSearchState by rememberUpdatedState(onResetSearch)
    val onResetUsersState by rememberUpdatedState(onResetUsers)

    // Observe tunables for reactivity
    val tunablesVersion = TunableConstants.version.value
    val buttonColor = TunableConstants.buttonColor()
    val buttonHighlightColor = TunableConstants.buttonHighlightColor()
    val cornerRoundness = TunableConstants.getDouble("corner-roundness", 1.0).toFloat()

    // Log toolbar appearance
    LaunchedEffect(Unit) {
        Logger.log("[Toolbar] Appeared")
    }

    // Register UI automation elements with updated state
    RegisterUIElement("toolbar-makepost") {
        if (currentExplorerState == ToolbarExplorer.MAKE_POST) {
            Logger.log("[Toolbar] Reset makePost view")
            onResetMakePostState()
        } else {
            Logger.log("[Toolbar] Explorer changed to: makePost")
            onExplorerChangeState(ToolbarExplorer.MAKE_POST)
        }
    }

    RegisterUIElement("toolbar-search") {
        if (currentExplorerState == ToolbarExplorer.SEARCH) {
            Logger.log("[Toolbar] Reset search view")
            onResetSearchState()
        } else {
            Logger.log("[Toolbar] Explorer changed to: search")
            onExplorerChangeState(ToolbarExplorer.SEARCH)
        }
    }

    RegisterUIElement("toolbar-users") {
        if (currentExplorerState == ToolbarExplorer.USERS) {
            Logger.log("[Toolbar] Reset users view")
            onResetUsersState()
        } else {
            Logger.log("[Toolbar] Explorer changed to: users")
            onExplorerChangeState(ToolbarExplorer.USERS)
        }
    }

    Box(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 8.dp)
            .padding(bottom = 34.dp)
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .shadow(
                    elevation = 12.dp,
                    shape = RoundedCornerShape((12 * cornerRoundness).dp),
                    ambientColor = Color.Black.copy(alpha = 0.4f),
                    spotColor = Color.Black.copy(alpha = 0.4f)
                )
                .clip(RoundedCornerShape((12 * cornerRoundness).dp))
                .background(buttonColor)
                .padding(horizontal = 66.dp, vertical = 16.dp),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            // Make Post button
            ToolbarButton(
                icon = Icons.Outlined.ChatBubbleOutline,
                isActive = currentExplorerState == ToolbarExplorer.MAKE_POST,
                showBadge = showPostsBadge,
                highlightColor = buttonHighlightColor,
                onClick = {
                    if (currentExplorerState == ToolbarExplorer.MAKE_POST) {
                        Logger.log("[Toolbar] Reset makePost view")
                        onResetMakePostState()
                    } else {
                        Logger.log("[Toolbar] Explorer changed to: makePost")
                        onExplorerChangeState(ToolbarExplorer.MAKE_POST)
                    }
                }
            )

            // Search button
            ToolbarButton(
                icon = Icons.Outlined.Search,
                isActive = currentExplorerState == ToolbarExplorer.SEARCH,
                showBadge = showSearchBadge,
                highlightColor = buttonHighlightColor,
                onClick = {
                    if (currentExplorerState == ToolbarExplorer.SEARCH) {
                        Logger.log("[Toolbar] Reset search view")
                        onResetSearchState()
                    } else {
                        Logger.log("[Toolbar] Explorer changed to: search")
                        onExplorerChangeState(ToolbarExplorer.SEARCH)
                    }
                }
            )

            // Users button
            ToolbarButton(
                icon = Icons.Outlined.People,
                isActive = currentExplorerState == ToolbarExplorer.USERS,
                showBadge = showUsersBadge,
                highlightColor = buttonHighlightColor,
                onClick = {
                    if (currentExplorerState == ToolbarExplorer.USERS) {
                        Logger.log("[Toolbar] Reset users view")
                        onResetUsersState()
                    } else {
                        Logger.log("[Toolbar] Explorer changed to: users")
                        onExplorerChangeState(ToolbarExplorer.USERS)
                    }
                }
            )
        }
    }
}

@Composable
fun ToolbarButton(
    icon: ImageVector,
    isActive: Boolean,
    showBadge: Boolean = false,
    highlightColor: Color = AppColors.accentHighlight,
    onClick: () -> Unit
) {
    val interactionSource = remember { MutableInteractionSource() }

    Box(
        modifier = Modifier
            .size(35.dp)
            .then(
                if (isActive) {
                    Modifier
                        .clip(RoundedCornerShape(6.dp))
                        .background(highlightColor)
                } else {
                    Modifier
                }
            )
            .clickable(
                interactionSource = interactionSource,
                indication = null,
                onClick = onClick
            ),
        contentAlignment = Alignment.Center
    ) {
        Icon(
            imageVector = icon,
            contentDescription = null,
            tint = Color.Black,
            modifier = Modifier.size(20.dp)
        )

        // Badge indicator
        if (showBadge) {
            Box(
                modifier = Modifier
                    .align(Alignment.TopEnd)
                    .offset(x = 2.dp, y = (-2).dp)
                    .size(10.dp)
                    .clip(CircleShape)
                    .background(Color.Red)
            )
        }
    }
}
